### PR TITLE
build: cleanup build scripts, use lazy APIs, update to Gradle 8.8.

### PR DESCRIPTION
**Pull request by autonomousapps**
Can be reviewed commit-by-commit, although two of the final commits are rather large (very samesy though).

### CODE REVIEW

File: gradle/libs.versions.toml [hunk 1]
```diff
@@ -112,23 +116,22 @@ jnrUnixsocket = { module = "com.github.jnr:jnr-unixsocket", version = "0.38.20"
...
-kotlinBom = { module = "org.jetbrains.kotlin:kotlin-bom", version = "1.9.21" }
-kotlinCompilerEmbeddable = { module = "org.jetbrains.kotlin:kotlin-compiler-embeddable", version = "1.9.21" }
-kotlinGradlePlugin = { module = "org.jetbrains.kotlin:kotlin-gradle-plugin", version = "1.9.21" }
+kotlinAllOpenPlugin = { module = "org.jetbrains.kotlin:kotlin-allopen", version.ref = "kotlin" }
+kotlinBom = { module = "org.jetbrains.kotlin:kotlin-bom", version.ref = "kotlin" }
+kotlinCompilerEmbeddable = { module = "org.jetbrains.kotlin:kotlin-compiler-embeddable", version.ref = "kotlin" }
+kotlinGradlePlugin = { module = "org.jetbrains.kotlin:kotlin-gradle-plugin", version.ref = "kotlin" }
 kotlinLogging = { module = "io.github.microutils:kotlin-logging", version = "3.0.5" }
-kotlinNoArgPlugin = { module = "org.jetbrains.kotlin:kotlin-noarg", version = "1.9.21" }
-kotlinReflect = { module = "org.jetbrains.kotlin:kotlin-reflect", version = "1.8.21" }
```
Comment (autonomousapps) [inline gradle/libs.versions.toml:128]: This appears to have been an error (using an older version of kotin-reflect).

File: gradle/libs.versions.toml [hunk 2]
```diff
@@ -112,23 +116,22 @@ jnrUnixsocket = { module = "com.github.jnr:jnr-unixsocket", version = "0.38.20"
 jooq = { module = "org.jooq:jooq", version = "3.18.2" }
 junitApi = { module = "org.junit.jupiter:junit-jupiter-api", version = "5.9.3" }
 junitEngine = { module = "org.junit.jupiter:junit-jupiter-engine", version = "5.9.3" }
-junitGradlePlugin = { module = "org.junit.platform:junit-platform-gradle-plugin", version = "1.2.0" }
```
Comment (autonomousapps) [inline gradle/libs.versions.toml:115]: This plugin doesn't do anything anymore.

File: misk-action-scopes/build.gradle.kts [hunk 3]
```diff
@@ -1,11 +1,9 @@
 import com.vanniktech.maven.publish.JavadocJar.Dokka
 import com.vanniktech.maven.publish.KotlinJvm
-import com.vanniktech.maven.publish.MavenPublishBaseExtension
 
 plugins {
-  kotlin("jvm")
-  `java-library`
```
Comment (autonomousapps) [inline misk-action-scopes/build.gradle.kts:7]: java-library is redundant when kotlin-jvm is applied.

File: misk-admin/build.gradle.kts [hunk 4]
```diff
@@ -49,7 +47,7 @@ sourceSets {
   }
 }
 
-val generatedSourceDir = "$buildDir/generated/source/wire-test"
```
Comment (autonomousapps) [inline misk-admin/build.gradle.kts:52]: `project.buildDir` is deprecated, to be removed in Gradle 9.

File: misk-bom/build.gradle.kts [hunk 5]
```diff
@@ -1,20 +1,20 @@
...
+  id("java-platform")
+  alias(libs.plugins.mavenPublishBase)
 }
 
 dependencies {
   constraints {
     project.parent?.subprojects?.forEach { subproject ->
       if (subproject.name != "misk-bom") {
-        api(subproject)
+        api(subproject.path)
```
Comment (autonomousapps) [inline misk-bom/build.gradle.kts:10]: There are occasionally weird issues when using a `project` reference directly. These two forms are equivalent.

File: misk-jooq/build.gradle.kts [hunk 6]
```diff
@@ -1,15 +1,23 @@
...
+
+// TODO(tsr): it might be better to put these on the root build classpath, but I'm not sure if that
+//  will have other repercussions. Do in a follow-up.
+// Needed to generate jooq test db classes
+buildscript {
+  dependencies {
+    classpath("org.flywaydb:flyway-gradle-plugin:9.14.1")
+    classpath(libs.mysql)
+  }
+}
```
Comment (autonomousapps) [inline misk-jooq/build.gradle.kts:12]: Note to self.

File: build.gradle.kts [hunk 7]
```diff
@@ -31,25 +29,15 @@ buildscript {
...
 
 plugins {
-  id("com.autonomousapps.dependency-analysis") version libs.versions.dependencyAnalysisPlugin.get()
-  id("org.jetbrains.kotlinx.binary-compatibility-validator") version libs.versions.kotlinBinaryCompatibilityPlugin.get()
-}
-
-apply(plugin = "com.vanniktech.maven.publish.base")
-
-allprojects {
-  group = when {
```
Comment (adrw) [inline build.gradle.kts:41]: Is this group stanza unnecessary?

**Review by adrw — APPROVED** [review]
Thanks for the cleanup!

File: build.gradle.kts (hunk 7 above)
Comment (autonomousapps) [reply build.gradle.kts:41]: I moved it into the settings script using a new API available in gradle 8.8!

File: settings.gradle.kts [hunk 8]
```diff
@@ -1,12 +1,37 @@
...
+  }
+}
+
+gradle.lifecycle.beforeProject {
+  group = when {
+    path.startsWith(":wisp") -> "app.cash.wisp"
+    else -> "com.squareup.misk"
   }
+  version = findProperty("VERSION_NAME") as? String ?: "0.0-SNAPSHOT"
 }
```
Comment (autonomousapps) [inline settings.gradle.kts:35]: New API in Gradle 8.8!

**Comment by autonomousapps** [comment]
api check failure:
```> API check failed for project misk-api.
  --- /home/runner/work/misk/misk/misk-api/api/misk-api.api
  +++ /home/runner/work/misk/misk/misk-api/build/api/misk-api.api
  @@ -56,7 +56,7 @@
   public abstract interface class misk/scope/ActionScoped {
   	public static final field Companion Lmisk/scope/ActionScoped$Companion;
   	public abstract fun get ()Ljava/lang/Object;
  -	public abstract fun getIfInScope ()Ljava/lang/Object;
  +	public fun getIfInScope ()Ljava/lang/Object;
   }```
   
should I just update it per this suggestion?
```   You can run :misk-api:apiDump task to overwrite API declarations
``

File: settings.gradle.kts (hunk 8 above)
Comment (yissachar) [reply settings.gradle.kts:35]: This is very cool

Comment (autonomousapps) [reply settings.gradle.kts:35]: tbh this is my first time using it and I'm not sure of all the implications -- hoping it is as good as it is ergonomic! This is meant to be isolated-projects-safe.

**Review by rainecp — APPROVED** [review]
Nice, thanks for simplifying this and making it more consistent!

File: build.gradle.kts [hunk 9]
```diff
@@ -251,27 +252,26 @@ subprojects {
...
+      dependsOn(tasks.named { it == "detektMain" })
     }
   }
 
+  val configurationNames = setOf("kapt", "wire", "proto", "Proto")
   configurations.all {
     // Workaround the Gradle bug resolving multiplatform dependencies.
     // https://github.com/square/okio/issues/647
-    if (name.contains("kapt") || name.contains("wire") || name.contains("proto") || name.contains("Proto")) {
+    if (name in configurationNames) {
```
Comment (staktrace) [inline build.gradle.kts:269]: This is not quite the same (`name.contains` vs `name ==`), but I don't know how relevant this copypasta is anymore anyway. Let's see if anything breaks!

File: misk-api/api/misk-api.api [hunk 10]
```diff
@@ -56,7 +56,7 @@ public abstract interface class misk/logging/Mdc {
 public abstract interface class misk/scope/ActionScoped {
 	public static final field Companion Lmisk/scope/ActionScoped$Companion;
 	public abstract fun get ()Ljava/lang/Object;
-	public abstract fun getIfInScope ()Ljava/lang/Object;
+	public fun getIfInScope ()Ljava/lang/Object;
```
Comment (staktrace) [inline misk-api/api/misk-api.api:59]: Seems like an odd change, do you know why this happened?

**Pull request merged**